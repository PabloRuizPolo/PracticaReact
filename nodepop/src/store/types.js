export const AUTH_LOGIN_PENDING = "auth/login/pending";
export const AUTH_LOGIN_COMPLETED = "auth/login/fulfilled";
export const AUTH_LOGIN_REJECTED = "auth/login/rejected";

export const AUTH_LOGOUT = "auth/logout";

export const ADDS_LOADED_PENDING = "adds/loaded/pending";
export const ADDS_LOADED_COMPLETED = "adds/loaded/fulfilled";
export const ADDS_LOADED_REJECTED = "adds/loaded/rejected";

export const ADDS_DETAIL = "adds/detail";

export const ADDS_CREATED_PENDING = "adds/created/pending";
export const ADDS_CREATED_COMPLETED = "adds/created/fulfilled";
export const ADDS_CREATED_REJECTED = "adds/created/rejected";

export const ADDS_DELETED_PENDING = "tweeaddsts/deleted/pending";
export const ADDS_DELETED_COMPLETED = "adds/deleted/fulfilled";
export const ADDS_DELETED_REJECTED = "adds/deleted/rejected";

export const UI_RESET_ERROR = "ui/reset-error";
