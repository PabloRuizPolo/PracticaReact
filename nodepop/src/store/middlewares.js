export function errorsMiddlewares(router, routesList) {
  return function (store) {
    return function (next) {
      return function (action) {
        const result = next(action);

        if (!action.error) {
          return result;
        }

        const go = routesList[action.payload.status];
        if (go) {
          router.navigate(go);
        }

        return result;
      };
    };
  };
}
