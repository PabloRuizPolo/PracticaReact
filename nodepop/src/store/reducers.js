import {
  ADDS_CREATED_COMPLETED,
  ADDS_DELETED_COMPLETED,
  ADDS_DETAIL,
  ADDS_DETAIL_COMPLETED,
  ADDS_LOADED_COMPLETED,
  AUTH_LOGIN_COMPLETED,
  AUTH_LOGOUT,
  UI_RESET_ERROR,
} from "./types";

export const defaultState = {
  auth: false,
  adds: {
    loaded: false,
    data: [],
  },
  ui: {
    pending: false,
    error: null,
  },
};

export function auth(state = defaultState.auth, action) {
  switch (action.type) {
    case AUTH_LOGIN_COMPLETED:
      return true;
    case AUTH_LOGOUT:
      return false;
    default:
      return state;
  }
}

export function adds(state = defaultState.adds, action) {
  switch (action.type) {
    case ADDS_LOADED_COMPLETED:
      return { ...state, loaded: true, data: action.payload };
    case ADDS_DETAIL:
      return { ...state, data: [action.payload] };
    case ADDS_CREATED_COMPLETED:
      return { ...state, data: [action.payload, ...state.data] };
    case ADDS_DELETED_COMPLETED:
      const filteredData = state.data.filter(
        (add) => add.id !== action.payload.id
      );
      return { ...state, data: filteredData };
    default:
      return state;
  }
}

export function ui(state = defaultState.ui, action) {
  if (action.error) {
    return { ...state, pending: false, error: action.payload };
  }
  if (action.type.endsWith("/pending")) {
    return { ...state, pending: true };
  }

  if (action.type.endsWith("/fulfilled")) {
    return { ...state, pending: false, error: null };
  }

  if (action.type === UI_RESET_ERROR) {
    return { ...state, error: null };
  }

  return state;
}
