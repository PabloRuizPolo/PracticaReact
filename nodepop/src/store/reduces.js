import {
  ADDS_LOADED_COMPLETED,
  ADDS_LOADED_PENDING,
  ADDS_LOADED_REJECTED,
  AUTH_LOGIN_COMPLETED,
  AUTH_LOGIN_PENDING,
  AUTH_LOGIN_REJECTED,
  AUTH_LOGOUT,
} from "./types";

export const defaultState = {
  auth: false,
  adds: {
    loaded: false,
    data: [],
  },
  error: null,
};

export const auth_login_pending = () => ({
  type: AUTH_LOGIN_PENDING,
});

export const auth_login_completed = () => ({
  type: AUTH_LOGIN_COMPLETED,
});

export const auth_login_rejected = (error) => ({
  type: AUTH_LOGIN_REJECTED,
  payload: error,
  error: true,
});

export const auth_logout = () => ({
  type: AUTH_LOGOUT,
});

export const adds_loaded_pending = () => ({
  type: ADDS_LOADED_PENDING,
});

export const adds_loaded_completed = (adds) => ({
  type: ADDS_LOADED_COMPLETED,
  payload: adds,
});

export const adds_loaded_rejected = (error) => ({
  type: ADDS_LOADED_REJECTED,
  payload: error,
  error: true,
});
