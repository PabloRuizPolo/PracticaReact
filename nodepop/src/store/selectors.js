export const getIsLogged = (state) => state.auth;

export const getAddsState = (state) => state.adds.data;
export const getAddLoaded = (state) => state.adds.loaded;

export const getAddState = (id) => (state) =>
  getAddsState(state).find((add) => add.id === id);

export const getTagsState = (state) => state.tags;

export const getUi = (state) => state.ui;
