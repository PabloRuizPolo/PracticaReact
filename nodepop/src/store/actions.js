import { getAddDetail, getAddLoaded, getAddState } from "./selectors";
import {
  ADDS_CREATED_COMPLETED,
  ADDS_CREATED_PENDING,
  ADDS_CREATED_REJECTED,
  ADDS_DELETED_COMPLETED,
  ADDS_DELETED_PENDING,
  ADDS_DETAIL,
  ADDS_LOADED_COMPLETED,
  ADDS_LOADED_PENDING,
  ADDS_LOADED_REJECTED,
  AUTH_LOGIN_COMPLETED,
  AUTH_LOGIN_PENDING,
  AUTH_LOGIN_REJECTED,
  AUTH_LOGOUT,
  UI_RESET_ERROR,
} from "./types";

export const auth_login_pending = () => ({
  type: AUTH_LOGIN_PENDING,
});
export const auth_login_completed = () => ({
  type: AUTH_LOGIN_COMPLETED,
});
export const auth_login_rejected = (error) => ({
  type: AUTH_LOGIN_REJECTED,
  payload: error,
  error: true,
});

export const authLogin = (inputValues) => {
  return async function (dispatch, _getState, { services: { auth } }) {
    try {
      dispatch(auth_login_pending());
      await auth.login(inputValues);
      dispatch(auth_login_completed());
    } catch (error) {
      dispatch(auth_login_rejected(error));
      throw error;
    }
  };
};

export const auth_logout = () => ({
  type: AUTH_LOGOUT,
});

export const adds_loaded_pending = () => ({
  type: ADDS_LOADED_PENDING,
});
export const adds_loaded_completed = (adds) => ({
  type: ADDS_LOADED_COMPLETED,
  payload: adds,
});
export const adds_loaded_rejected = (error) => ({
  type: ADDS_LOADED_REJECTED,
  payload: error,
  error: true,
});

export const addsLoad = () => {
  return async function (dispatch, getState, { services: { adds } }) {
    const state = getState();
    if (getAddLoaded(state)) {
      return;
    }
    try {
      dispatch(adds_loaded_pending());
      const addsPayload = await adds.getAdds();
      dispatch(adds_loaded_completed(addsPayload));
    } catch (error) {
      dispatch(adds_loaded_rejected(error));
      throw error;
    }
  };
};

export const addsDetail = (add) => ({
  type: ADDS_DETAIL,
  payload: add,
});

export const addLoad = (id) => {
  return async function (dispatch, getState, { services: { adds } }) {
    const state = getState();
    if (getAddState(id)(state)) {
      return;
    }

    const addPayload = await adds.getAdd(id);
    console.log(addPayload);
    dispatch(addsDetail(addPayload));
  };
};

export const adds_created_pending = () => ({
  type: ADDS_CREATED_PENDING,
});
export const adds_created_completed = (add) => ({
  type: ADDS_CREATED_COMPLETED,
  payload: add,
});
export const adds_created_rejected = (error) => ({
  type: ADDS_CREATED_REJECTED,
  payload: error,
  error: true,
});

export const adds_deleted_pending = () => ({
  type: ADDS_DELETED_PENDING,
});
export const adds_deleted_completed = (add) => ({
  type: ADDS_DELETED_COMPLETED,
  payload: add,
});
export const adds_deleted_rejected = (error) => ({
  type: ADDS_CREATED_REJECTED,
  payload: error,
  error: true,
});

export const uiResetError = () => ({
  type: UI_RESET_ERROR,
});
