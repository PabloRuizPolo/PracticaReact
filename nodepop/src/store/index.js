import { combineReducers, createStore, applyMiddleware } from "redux";
import { composeWithDevTools } from "@redux-devtools/extension";

import * as reducers from "./reducers";
import * as actions from "./actions";

const reducer = combineReducers(reducers);
const composeEnhancers = composeWithDevTools({ actions });

export default function confifureStore(preloadedState) {
  const store = createStore(
    reducer,
    preloadedState,
    composeEnhancers(
      applyMiddleware(withExtraArgument({ services: { auth, tweets } }))
    )
  );

  return store;
}
