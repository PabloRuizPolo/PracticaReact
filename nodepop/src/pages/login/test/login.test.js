import { render, screen } from "@testing-library/react";
import LoginPage from "../loginPage";
import { Provider } from "react-redux";
import userEvent from "@testing-library/user-event";
import { act } from "react";
import { authLogin } from "../../../store/actions";

const loginType = (input, text) => userEvent.type(input, text);
const loginClick = (input) => userEvent.click(input);
jest.mock("../../../store/actions");

describe("login", () => {
  const state = {
    ui: {
      pending: false,
      error: null,
    },
  };
  const store = {
    dispatch: () => {},
    subscribe: () => {},
    getState: () => state,
  };

  const renderComponents = () =>
    render(
      <Provider store={store}>
        <LoginPage />
      </Provider>
    );

  test("snapshot", () => {
    const { container } = renderComponents();
    expect(container).toMatchSnapshot();
  });

  test("should dispatch authLogin with credentials", () => {
    const email = "email";
    const password = "password";
    const checkbox = "false";
    renderComponents();

    const usernameInput = screen.getByLabelText("Email");
    const passwordInput = screen.getByLabelText("Contraseña");
    const checkboxInput = screen.getByLabelText("Recordar Contraseña");
    const loginButton = screen.getByRole("button");

    expect(loginButton).toBeDisabled();

    act(() => loginType(usernameInput, email));
    act(() => loginType(passwordInput, password));
    act(() => loginClick(checkboxInput));

    userEvent.click(loginButton);

    expect(authLogin).toHaveBeenCalledWith({
      email,
      password,
      checkbox,
    });
  });
});
