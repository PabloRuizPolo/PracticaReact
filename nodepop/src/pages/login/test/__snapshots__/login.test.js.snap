// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login snapshot 1`] = `
<div>
  <div
    class="loginPage"
  >
    <h2
      class="title"
    >
      Inicia sesión en Nodepop
    </h2>
    <form
      class="loginForm"
    >
      <div
        class="formField"
      >
        <label
          class="formField-label"
        >
          <span>
            Email
          </span>
          <input
            class="formField-input"
            name="email"
            type="mail"
            value=""
          />
        </label>
      </div>
      <div
        class="formField"
      >
        <label
          class="formField-label"
        >
          <span>
            Contraseña
          </span>
          <input
            class="formField-input"
            name="password"
            type="passqord"
            value=""
          />
        </label>
      </div>
      <div
        class="formField"
      >
        <label
          class="formField-label"
        >
          <span>
            Recordar Contraseña
          </span>
          <input
            class="formField-input"
            name="checkbox"
            type="checkbox"
            value="false"
          />
        </label>
      </div>
      <div
        class="div-button"
      >
        <button
          class="sc-beqWNU eKcPEv"
          disabled=""
          type="submit"
        >
          Entrar
        </button>
      </div>
    </form>
  </div>
</div>
`;
